homeassistant/components/switcher_kis/sensor.py

```diff
@@ -0,0 +1,131 @@
+"""Switcher integration Sensor platform."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from aioswitcher.consts import WAITING_TEXT
+from aioswitcher.devices import SwitcherV2Device
+
+from homeassistant.components.sensor import (
+    DEVICE_CLASS_CURRENT,
+    DEVICE_CLASS_POWER,
+    STATE_CLASS_MEASUREMENT,
+    SensorEntity,
+)
+from homeassistant.const import ELECTRICAL_CURRENT_AMPERE, POWER_WATT
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import DiscoveryInfoType, StateType
+
+from .const import DATA_DEVICE, DOMAIN, SIGNAL_SWITCHER_DEVICE_UPDATE
+
+
+@dataclass
+class AttributeDescription:
+    """Class to describe a sensor."""
+
+    name: str
+    icon: str | None = None
+    unit: str | None = None
+    device_class: str | None = None
+    state_class: str | None = None
+    default_enabled: bool = True
+    default_value: float | int | str | None = None
+
+
+POWER_SENSORS = {
+    "power_consumption": AttributeDescription(
+        name="Power Consumption",
+        unit=POWER_WATT,
+        device_class=DEVICE_CLASS_POWER,
+        state_class=STATE_CLASS_MEASUREMENT,
+        default_value=0,
+    ),
+    "electric_current": AttributeDescription(
+        name="Electric Current",
+        unit=ELECTRICAL_CURRENT_AMPERE,
+        device_class=DEVICE_CLASS_CURRENT,
+        state_class=STATE_CLASS_MEASUREMENT,
+        default_value=0.0,
+    ),
+}
+
+TIME_SENSORS = {
+    "remaining_time": AttributeDescription(
+        name="Remaining Time",
```

Comment:
This sounds like a relative time value. We don't want to store relative time in the state machine since that risks spamming it with changes. Time sensors should preferably be timestamp sensors with that device class.

Reply:
This is a counter which represents the time in minutes until the switch will turn off, the timer can either be set on the device panel or via a service: [Turn on with timer service](https://www.home-assistant.io/integrations/switcher_kis/#service-switcher_kisturn_on_with_timer)
Currently it will report the time left as a timestamp like this for 45 minutes it will report `00:45:00`, once the timer reach zero it will stay on the value of `00:00:00`, so unlike uptime sensors which will change forever, this sensor have a limited number of states reported and the value will only change when the device is turned on (when it is on the power sensors are sending updates every 4 seconds which is much higher rate than this sensor).

If I had designed it from start I would just report an integer with the minutes left and not in `hh:mm:ss` format, I assume this format was chosen since this is what displayed by the manufacture app.

I have tried converting it to a timestamp, it will work, but will have an odd behavior, when the timer reach zero the frontend will start showing as it is in the past. I have seen another implementation which just reports `None` when the timer is zero so the sensor will become `unknown`, but that doesn't look good either, in addition as a timestamp sensor the resolution of the sensor is not good enough if the timer is set to 90 minutes it will just show `In 1 hour`.

The whole reason for this PR is to migrate the switch properties to sensors as a preparation for a rewrite of the integration in order to support multiple devices via config flow discovery, which is a breaking change already, but I think that changing the behavior of this sensor will not be appreciated.

However if it is needed I will change it, as I see it there are 4 options:
- Normal timestamp, when timer is 0 the time will be in the past
- Timestamp, but when timer is 0 report `None` to mark the sensor as unknown
- Report the time left in minutes as an integer
- Leave it as it is now, a string formatted as `hh:mm:ss` 

What do you think?

Reply:
If the time resolution is minutes it's OK to keep it.